src/microbe_stage/PatchMap.cs

```diff
@@ -320,4 +370,127 @@ public void ReplaceSpecies(Species old, Species newSpecies)
             patch.ReplaceSpecies(old, newSpecies);
         }
     }
+
+    public void SetSeed(Random random)
+    {
+        if (Seed != null)
+            throw new ArgumentException("cant set seed, seed has already been set");
+
+        Seed = random;
+    }
+
+    public void BuildSpecialRegions()
+    {
+        foreach (var region in SpecialRegions)
+        {
+            region.Value.BuildRegion();
+        }
+    }
+
+    public void BuildPatchesInRegions()
+    {
+        foreach (var region in Regions)
+        {
+            region.Value.BuildPatches(Seed);
+            foreach (var patch in region.Value.Patches)
+            {
+                AddPatch(patch);
+            }
+        }
+    }
+
+    public void BuildPatchesInSpecialRegions()
+    {
+        foreach (var region in SpecialRegions)
+        {
+            region.Value.BuildPatches(Seed);
+            foreach (var patch in region.Value.Patches)
+            {
+                AddPatch(patch);
+            }
+        }
+    }
+
+    public void ConnectPatchesBetweenRegions()
+    {
+        foreach (var region in Regions)
+        {
+            region.Value.ConnectPatchesBetweenRegions(Seed);
+        }
+    }
+
+    public bool ContainsPatchAdjacency(int id1, int id2)
+    {
+        if (PatchesAdjacencies.Contains((id1, id2)) || PatchesAdjacencies.Contains((id2, id1)))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    public bool ContainsRegionAdjacency(int id1, int id2)
+    {
+        if (RegionAdjacencies.Contains((id1, id2)) || RegionAdjacencies.Contains((id2, id1)))
+        {
+            return true;
+        }
+
+        return false;
+    }
+
+    public void CreateAdjacencies()
+    {
+        foreach (var entry in Patches)
+        {
+            foreach (var adjacent in entry.Value.Adjacent)
+            {
+                if (!ContainsPatchAdjacency(entry.Value.ID, adjacent.ID))
+                    PatchesAdjacencies.Add((entry.Value.ID, adjacent.ID));
+            }
+        }
+
+        foreach (var entry in Regions)
+        {
+            foreach (var adjacent in entry.Value.Adjacent)
+            {
+                if (!ContainsRegionAdjacency(entry.Value.ID, adjacent.ID))
+                    RegionAdjacencies.Add((entry.Value.ID, adjacent.ID));
+            }
+        }
+    }
+
+    public void RecreateAdjacencies()
+    {
+        foreach (var (id1, id2) in PatchesAdjacencies)
+        {
+            var patch1 = Patches[id1];
+            var patch2 = Patches[id2];
+            patch1.AddNeighbour(patch2);
+            patch2.AddNeighbour(patch1);
+        }
+
+        foreach (var (id1, id2) in RegionAdjacencies)
+        {
+            PatchRegion region1;
+            if (id1 > -1)
```

Comment:
Why can an id be less than -1? This seems really arbitrary to me...

Reply:
I think Kemikal1 made `SpecialRegions`' id negative.

Reply:
I still don't even know what's so "special" about those regions. Unless they are *really* special I'll just recommend making them normal regions with a bool in them saying they are special or a more accurate name to whatever "special" really represents in a region.

Reply:
If they are just drawing subdivisions, I think it would make to sense model things with the region class so that it can have sub groupings in it. That way there doesn't need to be this extra complexity in the patch map class.

Reply:
This will add complexity to other parts of the code where you need to search each region for its subgroupings.

Reply:
Using negative id values is anyway an extra complication. I really don't like that, so pick either the sub regions approach or make all ids > 0.

Reply:
I dont know how to do it with ids > 0 and the ones with the subregions im not really in favor of or gonna do.